mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorRewriting.cpp

```diff
@@ -952,8 +952,15 @@ struct ReshapeRewriter : public OpRewritePattern<ReshapeOp> {
       auto rtp = getRankedTensorType(op.getResult());
       auto denseTp =
           RankedTensorType::get(rtp.getShape(), rtp.getElementType());
-      auto reshape = rewriter.create<ReshapeOp>(loc, denseTp, op.getSrc(),
-                                                op.getReassociation());
+      ReshapeOp reshape;
+      if constexpr (std::is_same<ReshapeOp, tensor::ExpandShapeOp>::value) {
```

Comment:
can this be hidden inside a templated "create" helper method, so we don't get such a huge if-then-else block?

Reply:
There are issues with templated helper. There are few more things which has to be handled for this patch like fusion, I can raise a PR with this and other issues fixed. It has lots of changes, so trying to not hold it for longer. Thanks :)

Reply:
+1, i have a bunch of things I want to do that depend on this, thanks for deferring improvements to follow-up PRs!

I'm also available to help with C++ templates if needed.